Common/Securities/CurrencyConversion/SecurityCurrencyConversion.cs

```diff
@@ -0,0 +1,283 @@
+        private SecurityCurrencyConversion(string sourceCurrency, string destinationCurrency, List<Step> steps)
+        {
+            SourceCurrency = sourceCurrency;
+            DestinationCurrency = destinationCurrency;
+
+            _steps = steps;
+        }
+
+        /// <summary>
+        /// Updates the internal conversion rate based on the latest data, and returns the new conversion rate
+        /// </summary>
+        /// <returns>The new conversion rate</returns>
+        public decimal Update()
+        {
+            var newConversionRate = 1m;
+            var stepWithoutDataFound = false;
+
+            _steps.ForEach(step =>
+            {
+                if (stepWithoutDataFound)
+                {
+                    return;
+                }
+
+                var lastData = step.RateSecurity.GetLastData();
+                if (lastData == null || lastData.Price == 0m)
+                {
+                    newConversionRate = 0m;
+                    stepWithoutDataFound = true;
+                    return;
+                }
+
+                if (step.Inverted)
+                {
+                    newConversionRate /= lastData.Price;
+                }
+                else
+                {
+                    newConversionRate *= lastData.Price;
+                }
+            });
+
+            ConversionRate = newConversionRate;
+            return ConversionRate;
+        }
+
+        /// <summary>
+        /// Finds a conversion between two currencies by looking through all available 1 and 2-leg options
+        /// </summary>
+        /// <param name="sourceCurrency">The currency to convert from</param>
+        /// <param name="destinationCurrency">The currency to convert to</param>
+        /// <param name="existingSecurities">The securities which are already added to the algorithm</param>
+        /// <param name="potentialSymbols">The symbols to consider, may overlap with existingSecurities</param>
+        /// <param name="makeNewSecurity">The function to call when a symbol becomes part of the conversion, must return the security that will provide price data about the symbol</param>
+        /// <returns>A new <see cref="SecurityCurrencyConversion"/> instance representing the conversion from sourceCurrency to destinationCurrency</returns>
+        /// <exception cref="ArgumentException">Thrown when no conversion from sourceCurrency to destinationCurrency can be found</exception>
+        public static SecurityCurrencyConversion LinearSearch(
+            string sourceCurrency,
+            string destinationCurrency,
+            IList<Security> existingSecurities,
+            IEnumerable<Symbol> potentialSymbols,
+            Func<Symbol, Security> makeNewSecurity)
+        {
+            var allSymbols = existingSecurities.Select(sec => sec.Symbol).Concat(potentialSymbols)
+                .Where(CurrencyPairUtil.IsDecomposable);
```

Comment:
nit - should ToList() this result since we might iterate up to 3 times through it bellow